server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java

```diff
@@ -349,9 +353,76 @@ public Repository repository(String repositoryName) {
         if (repository != null) {
             return repository;
         }
+        repository = internalRepositories.get(repositoryName);
+        if (repository != null) {
+            return repository;
+        }
         throw new RepositoryMissingException(repositoryName);
     }
 
+    public void registerInternalRepository(String name, String type, Settings settings) {
+        RepositoryMetaData metaData = new RepositoryMetaData(name, type, settings);
+        Repository existingRepository = internalRepositories.get(name);
+        if (existingRepository != null && existingRepository.getMetadata().equals(metaData)) {
+            // Identical repository already exists, no need to update it.
+            return;
+        }
+
+        // TODO: Normally we would do validation when we update a repository to ensure that it is not in use.
+        //  Are we okay with not including that validation under the assumption that internal operations
+        //  will do the right thing.
+
+        Repository newRepository = createRepository(metaData, internalTypesRegistry);
+        Repository repositoryToClose = null;
+        boolean updated = false;
+        synchronized (internalRepositories) {
```

Comment:
the concurrency logic looks complicated here. Maybe just add synchronized on the registerInternalRepository and unregisterInternalRepository methods? We will not be calling those concurrently anyway?

Reply:
After the changes where we do not support updates, we can rely on the `ConcurrentMap` to provide concurrency control.